### PR TITLE
Fix issue: wrong teamd link watch state after warm reboot

### PR DESCRIPTION
**Pull request by stephenxs**
#### Why I did it

Fix issue: wrong teamd link watch state after warm reboot due to TEAM_ATTR_PORT_CHANGED lost

The flag TEAM_ATTR_PORT_CHANGED is maintained by kernel team driver:
- a flag "changed" is maintained in struct team_port struct
- the flag is set by __team_port_change_send once relevant information is updated, including port linkup (together with speed, duplex), adding or removing
- the flag is cleared by team_nl_fill_one_port_get once the updated information has been notified to user space via RTNL

In the userspace, the change flag is maintained by libteam in struct team_port.
The team daemon calls port_priv_change_handler_func on receiving port change event.
The logic in port_priv_change_handler_func
1. creates the port if it did not exist, which triggers port add event and eventually calls lacp_port_added callback.
2. triggers port change event if team_port->changed is true, which eventually calls lw_ethtool_event_watch_port_changed to update port state for link watch ethtool.
3. removes the port if team_port->removed is removed

In lacp_port_added, it calls team_refresh to refresh ifinfo, port info, and option info from the kernel via RTNL.
In this step, port_priv_change_handler_func is called recursively.
- In the inner call, it won't get TEAM_ATTR_PORT_CHANGED flag because kernel has already notified that.
- As a result, team_port->changed flag is cleared in the libteam.
- The port change event won't be triggered from either inner or outer call of port_priv_change_handler_func.

If the port has been up when the port is being added to the team device, the "port up" information is carried in the outer call but will be lost.

In case the flag TEAM_ATTR_PORT_CHANGED is set only in the inner call, function port_priv_change_handler_func can be called in the inner call.
However, it will fail to fetch "enable" options because option_list_init has not be called.

Signed-off-by: Stephen Sun <stephens@nvidia.com>

#### How I did it

Fix:
Do not call check_call_change_handlers when parsing RTNL function is called from another check_call_change_handlers recursively.

#### How to verify it

- Manually test
- Regression test
  - warm reboot
  - warm reboot sad lag
  - warm reboot sad lag member
  - warm reboot sad (partial)
#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211

### CODE REVIEW

**Comment by stephenxs** [comment]
ci 2310 + rerun 275 passed

File: src/libteam/patch/0013-Fix-issue-TEAM_ATTR_PORT_CHANGED-can-be-lost-if-port.patch [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+@@ -128,6 +128,12 @@ int get_port_list_handler(struct nl_msg *msg, void *arg)
+ 	struct nlattr *port_attrs[TEAM_ATTR_PORT_MAX + 1];
+ 	int i;
+ 	uint32_t team_ifindex = 0;
++	/*
++	 * In case get_port_list is being called from check_call_change_handlers recursively,
++	 * there can be some attributes which have not been consumed by callbacks.
++	 * In this case, we should only merge the new attributes into the existing ones without clearing them
++	 */
++	bool recursive = (th->change_handler.pending_type_mask & TEAM_PORT_CHANGE) ? true : false;
```
Comment (stepanblyschak) [inline src/libteam/patch/0013-Fix-issue-TEAM_ATTR_PORT_CHANGED-can-be-lost-if-port.patch]: How does this indicate the call is recursive?

File: src/libteam/patch/0013-Fix-issue-TEAM_ATTR_PORT_CHANGED-can-be-lost-if-port.patch [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
++			port_list_cleanup_last_state(th);
+ 		th->msg_recv_started = true;
+ 	}
+ 	nla_for_each_nested(nl_port, attrs[TEAM_ATTR_LIST_PORT], i) {
+@@ -165,7 +172,8 @@ int get_port_list_handler(struct nl_msg *msg, void *arg)
+ 			if (!port)
+ 				return NL_SKIP;
+ 		}
+-		port->changed = port_attrs[TEAM_ATTR_PORT_CHANGED] ? true : false;
++                if (!port->changed || !recursive)
```
Comment (stepanblyschak) [inline src/libteam/patch/0013-Fix-issue-TEAM_ATTR_PORT_CHANGED-can-be-lost-if-port.patch]: Looks like there is tab/space inconsistency

**Review by stepanblyschak — COMMENTED** [review]
Is this is a fix for an issue that hapeens due to warm reboot patch? I suggest changing warm reboot patch instead of adding new one in that case.

I understand the recursion happens because of team_refresh in lacp_port_added? Why do we need team_refresh there?

File: src/libteam/patch/0013-Fix-issue-TEAM_ATTR_PORT_CHANGED-can-be-lost-if-port.patch (hunk 1 above)
Comment (stephenxs) [reply src/libteam/patch/0013-Fix-issue-TEAM_ATTR_PORT_CHANGED-can-be-lost-if-port.patch]: The logic is:
1. `TEAM_PORT_CHANGE` is set when handling the port change event in `get_port_list_handler`
2. after that, it calls `check_call_change_handlers(th, TEAM_PORT_CHANGE)`

In `check_call_change_handlers`, it
1. calls all handlers, including `port_priv_change_handler_func`
2. clears `TEAM_PORT_CHANGE` at the end of `check_call_change_handlers`

From the flow, we see that `TEAM_PORT_CHANGE` indicates the function is called from another handler, or recursively called.

**Comment by stephenxs** [comment]
> Is this is a fix for an issue that hapeens due to warm reboot patch? I suggest changing warm reboot patch instead of adding new one in that case.
> 
> I understand the recursion happens because of team_refresh in lacp_port_added? Why do we need team_refresh there?

Regarding merging it into the warm reboot patch, I'm OK only if it's a common practice to merge multiple patches into one. Regarding whether `team_refresh` is required in `lacp_port_added`, we need to further check it, but it was there from the beginning.

**Comment by stepanblyschak** [comment]
@stephenxs I believe so, since it is a fix for a patch that we introduced, it is common to change the original patch (or we can lose track of which patches are backports, which were developed by us) - e.g we already did it in the past -
```
- https://github.com/sonic-net/sonic-buildimage/pull/3544
- https://github.com/sonic-net/sonic-buildimage/pull/8227
```

I am Ok with these changes but I would like to better understand why team_refresh is being called from lacp_port_added. I am Ok to approve to discuss it with MSFT.

**Comment by stephenxs** [comment]
> @stephenxs I believe so, since it is a fix for a patch that we introduced, it is common to change the original patch (or we can lose track of which patches are backports, which were developed by us) - e.g we already did it in the past -
> 
> ```
> - https://github.com/sonic-net/sonic-buildimage/pull/3544
> - https://github.com/sonic-net/sonic-buildimage/pull/8227
> ```
> 
> I am Ok with these changes but I would like to better understand why team_refresh is being called from lacp_port_added. I am Ok to approve to discuss it with MSFT.

Thanks. I'm still checking whether there is a negative effect if `team_refresh` is removed. So far I didn't see any but still need to check more.
Regarding merging patches, I will do.

**Comment by stephenxs** [comment]
> @stephenxs I believe so, since it is a fix for a patch that we introduced, it is common to change the original patch (or we can lose track of which patches are backports, which were developed by us) - e.g we already did it in the past -
> 
> ```
> - https://github.com/sonic-net/sonic-buildimage/pull/3544
> - https://github.com/sonic-net/sonic-buildimage/pull/8227
> ```
> 
> I am Ok with these changes but I would like to better understand why team_refresh is being called from lacp_port_added. I am Ok to approve to discuss it with MSFT.

Hi @stepanblyschak 
I do not see any solid reason that `team_refresh` is required to be called in `lacp_port_added`.
One possible reason is that `lacp_port_added` calls `team_set_port_enabled` which requires options and if the option is not available at the time (eg. hasn't received kernel notification yet), it can accelerate the process by calling `team_refresh`. But I'm not sure whether it is necessary because in very short time it should receive the notification from the kernel.
So, I'm also open to other options, eg. to remove `team_refresh` from `lacp_port_added`.
I think we can discuss the current solution with MSFT to test the water.

btw, I merged the patch to the original one.

**Review by stepanblyschak — APPROVED** [review]

**Comment by stephenxs** [comment]
ci 2388 + rerun 313 passed

**Pull request closed without merging**